### PR TITLE
Fix on message registration delay issue

### PR DESCRIPTION
**Pull request by vinaysurya**
## Description
Address issue #189 
Do not block waiting for a message/session when Registering a message/session handler.
Also make the exception handling contract a must when registering an on OnMessage/OnSession handler() so that user thinks about exceptions that might happen on the OnMessage Pump right from the getgo. 

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] **I have read the [contribution guidelines](./CONTRIBUTING.md).**
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR.
- [x] The pull request does not introduce breaking changes (unless a major version change occurs in the assembly and module).
- [x] If applicable, the public code is properly documented.
- [x] Pull request includes test coverage for the included changes.
- [x] The code builds without any errors.

### CODE REVIEW

File: src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs [hunk 1]
```diff
@@ -32,15 +32,9 @@ sealed class MessageReceivePump
...
+        public void StartPump()
         {
-            var initialMessage = await this.messageReceiver.ReceiveAsync(Constants.MessageReceiverStartPumpInitialReceiveTimeout).ConfigureAwait(false);
-            if (initialMessage != null)
-            {
-                MessagingEventSource.Log.MessageReceiverPumpInitialMessageReceived(this.messageReceiver.ClientId, initialMessage);
-            }
-
-            TaskExtensionHelper.Schedule(() => this.MessagePumpTask(initialMessage));
+            TaskExtensionHelper.Schedule(() => this.MessagePumpTaskAsync());
```
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs]: With #186 in place, is this the right thing to do?

File: src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs [hunk 2]
```diff
@@ -1,32 +1,37 @@
...
+        public string Action { get; private set; }
+
+        /// <summary>The namespace name used when this exception occured.</summary>
+        public string Endpoint { get; private set; }
+
+        /// <summary>The entity path used when this exception occured.</summary>
+        public string EntityPath { get; private set; }
+
+        /// <summary>The Client Id associated with the sender, receiver or session when this exception occured.</summary>
+        public string ClientId { get; private set; }
```
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs]: Client ID is used for object identification, right?
By reading xmldoc I would not understand why this is exposed and what would I do with it when exception is thrown.

File: src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs [hunk 3]
```diff
@@ -93,10 +72,10 @@ bool ShouldRenewSessionLock()
                 this.sessionHandlerOptions.AutoRenewLock;
         }
 
-        void RaiseExceptionReceived(Exception e, string action)
+        async Task RaiseExceptionReceived(Exception e, string action)
         {
-            var eventArgs = new ExceptionReceivedEventArgs(e, action, this.endpoint, this.entityPath);
-            this.sessionHandlerOptions.RaiseExceptionReceived(eventArgs);
+            var eventArgs = new ExceptionReceivedEventArgs(e, action, this.endpoint, this.entityPath, this.clientId);
+            await this.sessionHandlerOptions.RaiseExceptionReceived(eventArgs).ConfigureAwait(false);
```
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs]: It it's invoked and no need in the awaited result, method can drop `async` and return the task

`return this.sessionHandlerOptions.RaiseExceptionReceived(eventArgs)`

File: src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs (hunk 2 above)
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs]: Client ID is used for object identification, right?
By reading xmldoc I would not understand why this is exposed and what would I do with it when exception is thrown.

File: src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs (hunk 1 above)
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs]: With #186 in place, is this the right thing to do?

File: src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs (hunk 2 above)
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs]: Would rewording to "The Client Id can be used to associate the exception with the queueclient, subscriptionClient, MessageSender or MessageReceiver  that encountered the exception"

File: src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs (hunk 1 above)
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs]: I haven't spent time on 186, but since we are not changing that behavior here,  I would prefer to handle that as a separate issue itself.

Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs]: @SeanFeldman IMO, we should tackle this and #186 separately.

File: src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs (hunk 2 above)
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs]: 👍 
Still, in production, where let's say it's logged, how this piece of information can help? I might ask a stupid question, but this is what I'd ask when get this info in a log stream.

File: src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs (hunk 1 above)
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs]: OK

**Comment by vinaysurya** [comment]
Good point about the sample update. I've created a tracking issue on that 
https://github.com/Azure/azure-service-bus/issues/46

File: src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs (hunk 2 above)
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs]: Basically lets say you create multiple  Queueclients and want to figure out which of the queueclients are throwing that exception for you, then this identifier can be used to figure that out.

Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs]: Wouldn't that only be possible when you have debugger attached?

Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs]: If you are using logs, we log clientId's with most events right?

Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/Primitives/ExceptionReceivedContext.cs]: logs as in EventSource events i mean

**Comment by SeanFeldman** [comment]
BTW, https://github.com/blog/2389-requesting-reviews-from-repository-teams
